### PR TITLE
Purchases: added configuration errors for missing products

### PR DESCRIPTION
**Pull request by NachoSoto**
Fixes [sc-11103].

### CODE REVIEW

File: purchases/src/test/java/com/revenuecat/purchases/PurchasesTest.kt [hunk 1]
```diff
@@ -1011,27 +1073,6 @@ class PurchasesTest {
         }
     }
 
-    @Test
-    fun getOfferingsErrorIsNotCalledIfSkuDetailsMissing() {
```
Comment (NachoSoto) [inline purchases/src/test/java/com/revenuecat/purchases/PurchasesTest.kt:1015]: This test was incompatible with the new expected behavior, and with the behavior in iOS.
`configuration error if products are not set up` has the same setup, but checks for the existing error.

Let me know if I'm missing something?

File: purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt [hunk 2]
```diff
@@ -1088,6 +1097,22 @@ class Purchases @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE) intern
             })
     }
 
+    private fun extractSkus(offeringsJSON: JSONObject): Set<String> {
+        val jsonArrayOfOfferings = offeringsJSON.getJSONArray("offerings")
```
Comment (beylmk) [inline purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt:1101]: i've always wondered if it's worth pulling these json keys into constants somewhere. thoughts? probably depends on how often we use them

Comment (NachoSoto) [reply purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt:1101]: What does everyone think of using [`GSON`](https://github.com/google/gson) instead?

Comment (NachoSoto) [reply purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt:1101]: (By the way this method is just extracted from the existing implementation).

File: purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt [hunk 3]
```diff
@@ -1048,31 +1048,40 @@ class Purchases @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE) intern
...
-                            skus.add(
-                                jsonPackagesArray.getJSONObject(j)
-                                    .getString("platform_product_identifier")
-                            )
-                        }
+                    val skus = extractSkus(offeringsJSON)
+                    if (skus.isEmpty()) {
+                        handleErrorFetchingOfferings(
+                            PurchasesError(
+                                PurchasesErrorCode.ConfigurationError,
```
Comment (beylmk) [inline purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt:1055]: this reminds me, there was a request in a `purchases-ios` issue that we update our documentation. it currently says offerings aren't required, but then we error out. we should probably clarify why/when this error is expected. @codykerns thoughts?

File: purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt [hunk 4]
```diff
@@ -1048,31 +1048,40 @@ class Purchases @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE) intern
...
+                            if (offerings.all.isEmpty()) {
+                                handleErrorFetchingOfferings(
+                                    PurchasesError(
+                                        PurchasesErrorCode.ConfigurationError,
+                                        OfferingStrings.CONFIGURATION_ERROR_PRODUCTS_NOT_FOUND
+                                    ),
+                                    completion
+                                )
+                            } else {
+                                logMissingProducts(offerings, detailsByID)
```
Comment (beylmk) [inline purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt:1073]: should we log missing products when empty too? or somehow else pass in which products were not found?

File: purchases/src/test/java/com/revenuecat/purchases/PurchasesTest.kt [hunk 5]
```diff
@@ -913,6 +918,63 @@ class PurchasesTest {
         assertThat(receivedOfferings!![stubOfferingIdentifier]!!.monthly!!.product).isNotNull
     }
 
+    @Test
+    fun `configuration error if no products are found`() {
```
Comment (beylmk) [inline purchases/src/test/java/com/revenuecat/purchases/PurchasesTest.kt:922]: the current errors are logged with `LogIntent.GOOGLE_ERROR`, i wonder if we should be using something else...

File: purchases/src/test/java/com/revenuecat/purchases/PurchasesTest.kt [hunk 6]
```diff
@@ -125,6 +125,11 @@ class PurchasesTest {
         "{'identifier': '\$rc_monthly','platform_product_identifier': '$stubProductIdentifier'}" +
         "]}]," +
         "'current_offering_id': '$stubOfferingIdentifier'}"
+    private val noOfferingsResponse = "{'offerings': [" +
```
Comment (beylmk) [inline purchases/src/test/java/com/revenuecat/purchases/PurchasesTest.kt:128]: doesn't this have an offering in it? it has no packages...

File: purchases/src/test/java/com/revenuecat/purchases/PurchasesTest.kt (hunk 1 above)
Comment (beylmk) [reply purchases/src/test/java/com/revenuecat/purchases/PurchasesTest.kt:1015]: that makes sense to me..

File: purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt (hunk 2 above)
Comment (beylmk) [reply purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt:1101]: we've avoided adding any 3rd party libs.... @vegaro probably has more historical info on that though, and whether we might open up our thinking..

Comment (NachoSoto) [reply purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt:1101]: Oh that's right.

File: purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt (hunk 4 above)
Comment (NachoSoto) [reply purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt:1073]: Oh yeah that makes sense, we do that in iOS.

File: purchases/src/test/java/com/revenuecat/purchases/PurchasesTest.kt (hunk 5 above)
Comment (NachoSoto) [reply purchases/src/test/java/com/revenuecat/purchases/PurchasesTest.kt:922]: Oooh good catch. I added a proposed solution but let me know if we should do a less localized fix, maybe putting the logic in `PurchasesErrorCode`.

File: purchases/src/test/java/com/revenuecat/purchases/PurchasesTest.kt (hunk 6 above)
Comment (NachoSoto) [reply purchases/src/test/java/com/revenuecat/purchases/PurchasesTest.kt:128]: Fixed, thanks.

File: purchases/src/test/java/com/revenuecat/purchases/PurchasesTest.kt (hunk 5 above)
Comment (beylmk) [reply purchases/src/test/java/com/revenuecat/purchases/PurchasesTest.kt:921]: nice 👍  i think what you added is good for now... i think we have some error cleanup to do on both natives that probably needs some discussion anyway.

File: purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt (hunk 2 above)
Comment (vegaro) [reply purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt:1102]: We should use Kotlin Serialization https://kotlinlang.org/docs/serialization.html#example-json-serialization

I thought we had a story for it but I couldn't find it so I created one https://app.shortcut.com/revenuecat/story/12198/use-kotlin-serialization-instead-of-manual-serialization

**Comment by NachoSoto** [comment]
Anything else pending here?

**Review by beylmk — APPROVED** [review]
🎊

**Pull request merged**